Source/WebCore/platform/graphics/cocoa/UnrealizedCoreTextFont.cpp

```diff
@@ -334,6 +335,114 @@ void UnrealizedCoreTextFont::modifyFromContext(const FontDescription& fontDescri
+            return { };
+        return result;
+    }
+    return { };
+}
+#endif
+
+// Once rdar://problem/105483251 and rdar://problem/108877507 are solved, we can delete this function and the whole font rebuilding codepath.
+auto UnrealizedCoreTextFont::rebuildReason(CTFontRef font) const -> RebuildReason
+{
+    RebuildReason rebuildReason;
+
+    // Both OpenType 1.8 and TrueType GX support font variations, using the same named keys and values.
+    // However, the scale is different between the two: OpenType uses the CSS scale, whereas TrueType
+    // uses a 0-1-2 scale.
+    //
+    // However, we don't actually know whether the font is a TrueType GX font or not until we create a
+    // CTFont (which is done just above). The vast majority of fonts are OpenType fonts, so we
+    // opportunistically assume the font at hand is one of those and set the values accordingly. Now,
+    // after we've created the font, we can check to see whether or not our guess was correct. If it
+    // wasn't, we have to recreate the font with the normalized values.
+    if (m_applyTraitsVariations == ApplyTraitsVariations::Yes && FontInterrogation(font).variationType == FontInterrogation::VariationType::TrueTypeGX) {
+        auto variationValues = defaultVariationValues(font, ShouldLocalizeAxisNames::No);
+        if (variationValues.contains({ { 'w', 'g', 'h', 't' } })
+            || variationValues.contains({ { 'w', 'd', 't', 'h' } })
+            || variationValues.contains({ { 'i', 't', 'a', 'l' } })
+            || variationValues.contains({ { 's', 'l', 'n', 't' } }))
+            rebuildReason.gxVariations = true;
+    }
+
+    // When the fix for rdar://problem/108877507 is deployed everywhere, we can delete this.
+#if USE(CORE_TEXT_OPTICAL_SIZING_RANGE_CLAMPING_WORKAROUND)
+    if (auto range = opticalSizeRange(font)) {
+        WTF::switchOn(m_opticalSizingType, [&](OpticalSizingTypes::None) {
+        }, [&](OpticalSizingTypes::JustVariation) {
+            if (range->isInRange(m_size))
+                return;
+            rebuildReason.outOfRangeVariations = std::clamp(m_size, range->minimum, range->maximum);
+        }, [&](const OpticalSizingTypes::Everything& everything) {
+            if (!everything.opticalSizingValue)
+                return;
+            if (range->isInRange(m_size))
+                return;
+            rebuildReason.outOfRangeVariations = std::clamp(*everything.opticalSizingValue, static_cast<float>(range->minimum), static_cast<float>(range->maximum));
+        });
+        auto iterator = std::find_if(m_variationSettings.begin(), m_variationSettings.end(), [&](const auto& setting) {
+            return setting.tag() == FontTag { { 'o', 'p', 's', 'z' } };
+        });
+        if (iterator != m_variationSettings.end())
+            rebuildReason.outOfRangeVariations = std::clamp(iterator->value(), static_cast<float>(range->minimum), static_cast<float>(range->maximum));
+    }
+#endif
```

Comment:
can't we guard this fragment against `FontDescription::opticalSizing()` such that we avoid querying CT if optical sizing is ::Disabled?

Reply:
That's what `OpticalSizingTypes::None` does. If optical sizing is disabled, then `m_opticalSizingType` is set to `OpticalSizingTypes::None`.